pkg/odo/cli/list/list.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// Complete ...
+func (lo *ListOptions) Complete(cmdline cmdline.Cmdline, args []string) (err error) {
+
+	// Check to see if KUBECONFIG exists, and if not, error the user that we would not be able to get cluster information
+	// Do this before anything else, or else we will just error out with the:
+	// invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
+	// instead
+	if !dfutil.CheckKubeConfigExist() {
+		return errors.New("KUBECONFIG not found. Unable to retrieve cluster information. Please set your Kubernetes configuration via KUBECONFIG env variable or ~/.kube/config")
+	}
+
+	// Create the local context and initial Kubernetes client configuration
+	lo.Context, err = genericclioptions.New(genericclioptions.NewCreateParameters(cmdline))
+	if err != nil {
+		return err
+	}
+
+	// Check for the Devfile and then retrieve all information regarding the local Devfile
+	lo.devfilePath = location.DevfileLocation("")
+	if util.CheckPathExists(lo.devfilePath) {
+
+		// Set the project / namespace based on the devfile context
+		lo.project = lo.Context.GetProject()
+
+		// Parse the devfile
+		devObj, parseErr := devfile.ParseAndValidateFromFile(lo.devfilePath)
+		if parseErr != nil {
+			return parseErr
+		}
+
+		// Convert the devfile to a Component{} for table output and manipulation
+		// we want to limit using lo.EnvSpecificInfo, but fortunately, it will simply try to find
+		// lo.devfilePath for our usage.
+		localComponent, err := component.GetComponentFromDevfile(devObj, lo.project)
+		if err != nil {
+			return err
+		}
+		lo.localComponent = localComponent
+
+	}
+
+	// If the context is "" (devfile.yaml not found..), we get the active one from KUBECONFIG.
+	if lo.project == "" {
+		lo.project = lo.clientset.KubernetesClient.GetCurrentNamespace()
+	}
+
+	// If the namespace flag has been passed, we will search there.
+	// if it hasn't, we will search from the default project / namespace.
+	if lo.namespaceFlag != "" {
+		lo.namespaceFilter = lo.namespaceFlag
+	} else {
+		lo.namespaceFilter = lo.project
+	}
+
+	return
+}
+
+// Validate ...
+func (lo *ListOptions) Validate() (err error) {
+	return nil
+}
+
+// Run has the logic to perform the required actions as part of command
+func (lo *ListOptions) Run() error {
+
+	// Set the correct selector / project before continuing in case we have passed in
+	// --namespace
+
+	// All variables pertaining to output / combination
+	devfileComponents := []component.Component{}
+
+	// Step 1.
+	// Retrieve all related components from the Kubernetes cluster, from the given namespace
+	if lo.clientset.KubernetesClient != nil {
+
+		// Create a spinner since this function takes up 99% of `odo list` time
+		listSpinner := log.Spinnerf("Listing components from namespace '%s'", lo.namespaceFilter)
+		defer listSpinner.End(false)
+
+		devfileComponentsOut, err := component.ListAllClusterComponents(lo.clientset.KubernetesClient, lo.namespaceFilter)
```

Comment:
Why are you using a temp variable to store the result? If this is to handle the case when you get an err, you don't have to care about the value of `devfileComponents` as you return immediately